Make gap generators work when shroud is disabled.

deniz1a opened the pull request:
Fixes #6123. This will look at game settings, if fog of war is disabled, gap gens will work like they do now. If fog of war is enabled, gap gens will generate fog and there won't be any frozen actors visible while they are operational. Also this commit disables auto targeting for units/structures under gap, even if you have satellite. So satellite icons will also not appear where gap gen is active.

I tested this and seems to work.

RoosterDragon commented on OpenRA.Game/Orders/UnitOrderGenerator.cs:
```diff
@@ -74,7 +74,7 @@ public string GetCursor(World world, CPos xy, MouseInput mi)
 			{
 				var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mi)
-					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures))
+					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures) && !world.RenderPlayer.Shroud.IsUnderGapGen(a.Footprint))
```
This code is duplicated in the Order method, will that need updating too?

deniz1a replied:
I don't know, it seems to work with just this code. This is to prevent having an attack cursor for hidden buildings under fog (they dont have frozen images now if gap gen is active).

RoosterDragon commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -119,8 +119,8 @@ public FrozenActorLayer(Actor self)
-			frozen.Add(fa.ID, fa);
-			world.ScreenMap.Add(owner, fa);
+                frozen.Add(fa.ID, fa);
+                world.ScreenMap.Add(owner, fa);
```
Use tabs for indentation.

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -28,7 +28,7 @@ public class Shroud
 
 		readonly CellLayer<short> visibleCount;
-		readonly CellLayer<short> generatedShroudCount;
+		public readonly CellLayer<short> generatedGapCount;
```
You can keep this private.

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -275,7 +275,10 @@ public bool IsExplored(MPos uv)
+            if (FogEnabled)
+                return explored[uv];
+            else
+                return explored[uv] && (generatedGapCount[uv] == 0 || visibleCount[uv] > 0);
```
```return explored[uv] && (FogEnabled || generatedGapCount[uv] == 0 || visibleCount[uv] > 0);``` would be equivalent, also check indentation here.

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -359,5 +372,21 @@ public bool HasFogVisibility()
+        
+        public bool IsUnderGapGen(MPos[] fp)
+        {
+            if (fp.All(c => generatedGapCount[c] > 0))
```
You can do ```return fp.All(c => generatedGapCount[c] > 0);```. Check indentation.

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -359,5 +372,21 @@ public bool HasFogVisibility()
+                return false;
+        }
+        
+        public bool IsUnderGapGen(IEnumerable<CPos> fp)
```
You can do ```return fp.All(c => generatedGapCount[c] > 0);```. Check indentation.

RoosterDragon commented on OpenRA.Mods.RA/Effects/GpsDot.cs:
```diff
@@ -65,6 +65,9 @@ public GpsDot(Actor self, GpsDotInfo info)
 
 		bool ShouldShowIndicator()
 		{
+            if(self.World.RenderPlayer.Shroud.IsUnderGapGen(Shroud.GetVisOrigins(self)))
```
Check indentation and leave a space after the ```if``` keyword.

deniz1a commented:
OK I think I fixed all code styling issues. I also squashed all commits.

deniz1a commented:
Also for gap gen to cover a building, all its footprint cells must be within range. So if part of the building is outside gap gen range it will appear as normal. Does footprint include bibs, the space below structures?

pchote commented:
Shroud is an engine-level feature, so please don't introduce RA-specific logic and naming there.  You can do that instead in the mod-level traits.

deniz1a commented:
Do you mean UpdateShroudGeneration to UpdateGapGeneration etc changes? But it doesn't generate shroud anymore if fog is enabled. Isn't gap a general name for it?

Phrohdoh commented:
No, shroud is the name for shroud and fog is the name for fog.

deniz1a commented:
OK so how should it be called? It's not a shroud generator or fog generator but a combination of those. ShroudFogGeneration?

deniz1a commented:
Actually if fog is enabled, gap generator doesn't generate fog. The fog is generated automatically by the engine. What gap gen does then is remove frozen actors from fog. So it either generates shroud or removes frozen actors from fog. WHat should it be called? Gap generator seems to be the most appropriate name, it generates a gap in enemy radar/map. Are there any other name suggestions? ShroudFogGeneration doesn't look suitable since gap gen doesn't generate fog.

deniz1a commented:
Also the other point pchote made, I haven't introduced any ra specific code here? The shroud generation code was already there, I just changed its behaviour. And don't all mods have shroud and fog option? So it's not mod specific.

pchote commented:
The concept of having the "shroud generation" do one thing if shroud is enabled but another if it isn't is a RA specific idea.  Other mods may want to have pure shroud generators or pure cloak generators, but you are instead enforcing them to do what the gap generator does.

The shroud code needs to keep these ideas separate (with separate methods and state for shroud and cloak generation), and then the gameplay code (i.e. CreatesShroud) should expose trait properties so that the generation type can be configured in the yaml.

deniz1a commented:
OK it was agreed in IRC to call this functionality "Stealth Field Generation". I will make the necessary changes and send another commit.

deniz1a commented:
I decided to postpone this until #7577 is merged and #7669 is fixed.

deniz1a commented:
OK I have another naming question: So the general trait will be called CreatesStealth, it is a general stealth generation field. And then there will be two types of stealth. One is the current gap generator, it will generate shroud, so it's called type: shroud. But what should the new type be called? The new gap generator behaviour will act as frozen actor remover from fog and also prevent units from being auto targeted. Sensor disruptor was suggested in IRC. What do you think?

Phrohdoh commented:
:-1: for `CreatesStealth`, shroud is not stealth/cloak.

penev92 commented:
This is basically what I've gathered from all the discussions:
 - Working with Shroud disabled - it generates fog
 - Working with Fog disabled - it generates shroud
 - Working with both Shroud and Fog disabled - it just makes things "invisible" unless you are near them
 - Working with both Shroud and Fog enabled - generate fog

deniz1a commented:
Yes. Or simpler: if fog is enabled, gap gen acts as disruption generator. If fog is disabled but shroud is enabled, it generates shroud. If both are disabled, it makes units invisible?

deniz1a commented:
@Phrohdoh: So what should it be called instead of CreatesStealth.cs?

deniz1a committed:
Introduces a new trait called CreatesDisruptionField which makes Gap Gens work when shroud is disabled.

Performance optimization

Uses IsDisabled() method for simplification.

Optimizes fog/shroud checks.

Adds a new trait called RenderDisruptionCircle.

Converts some IEnumerables to array.

Simplifies if statements.

Uses ActorPreview for rendering.

Shortens some lines.

Reverts faulty project file changes.

Phrohdoh commented on OpenRA.Game/OpenRA.Game.csproj:
```diff
@@ -349,4 +349,4 @@
   -->
   <ItemGroup />
-</Project>
+</Project>
```
This needs to be reverted.

deniz1a replied:
What needs to be reverted? Those changes in the project file were done by Visual Studio automatically when I renamed CreatesShroud.cs to CreatesConcealment.cs.

RoosterDragon replied:
You should stage just the ```<Compile Include="Traits\CreatesDisruptionField.cs" />``` line. We're trying to avoid staging other lines the different IDEs keep automatically changing so they don't go back and forth all the time.

deniz1a commented:
OK I have updated the code. Another review please.

deniz1a commented:
I didn't add separate methods for disruption field generation, I couldn't decide which one would be better. If you say separate methods for disruption in Shroud.cs would be better, I can change it. And I made a single trait for both types. I think this is better than having two traits and then checking if the actor also has the other trait... Because they are basically the same trait with slightly different behaviour.

deniz1a commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -30,15 +30,17 @@ public class Shroud
-		readonly CellLayer<short> generatedShroudCount;
 		readonly CellLayer<bool> explored;
+		readonly CellLayer<short> generatedDisruptionCount;
+		readonly CellLayer<short> generatedShroudCount;
```
It doesn't work when I make this CellLayer<ushort>. Why? Does it ever fall to negative values?

Phrohdoh replied:
An unsigned type can **never** be negative.

deniz1a replied:
I know. But does it fall to negative when it's short?

Phrohdoh replied:
You tell us.
Add some debug lines and you'll figure it out.

deniz1a commented:
I have finished the code. It adds a new trait called CreatesDisruptionField. This makes gap gens work without shroud.

Phrohdoh commented:
This certainly *works*, but as I've stated before I'm not an advocate for this logic.

I do not think this is visually pleasing: (structures disappearing into the fog abruptly)
![gif](http://i.imgur.com/5QjbVRX.gif)

deniz1a commented:
OK but how could it be done otherwise so that gap generators work when shroud is disabled?

Phrohdoh commented on OpenRA.Game/Traits/CreatesDisruptionField.cs:
```diff
@@ -0,0 +1,51 @@
+{
+	public class CreatesDisruptionFieldInfo : ITraitInfo
+	{
+		public readonly WRange Range = WRange.Zero;
```
Add a `Desc` attribute to this, please.

Phrohdoh commented on OpenRA.Game/Traits/CreatesDisruptionField.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace OpenRA.Traits
+{
+	public class CreatesDisruptionFieldInfo : ITraitInfo
```
Add a `Desc` attribute to this, please.

Phrohdoh commented on OpenRA.Game/Traits/CreatesDisruptionField.cs:
```diff
@@ -0,0 +1,51 @@
+			}
+		}
+
+		public WRange Range { get { return cachedDisabled ? WRange.Zero : info.Range; } }
```
This should go above all methods and private members.

Phrohdoh commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -163,42 +166,105 @@ void UpdateVisibility(Actor a, ref CPos[] visible)
+				generatedDisruptionCount[c]++;
+
+			if (disruptionGenerators.ContainsKey(a))
+				throw new InvalidOperationException("Attempting to add duplicate shroud generation");
```
If you are going to change logic please update any accompanying messages.
This is now *disruption*, not shroud.

deniz1a commented:
This code is still not finished. I found some bugs and am trying to fix them...

deniz1a commented:
This patch has turned out to be more complex than I had thought. Currently it has the following problems:
1) If a previously discovered building that is under active gap gen (disruption field) is sold or destroyed, its frozen actor doesn't reappear if the gap gen is not active anymore. In other words, the information that it is no longer there is revealed, which is not done for normal frozen actors.
2) Whether a frozen actor is rendered or not is decided by using its footprint. But whether it can be auto-targeted is decided by its OccupiedCells(). These are different areas, which results in buildings that are visible as frozen actors but not auto-targeted, if the occupied cells are under disruption field but the bibs of the building (which are part of the footprint) are not.
3) When a move order is given over a hidden frozen actor, it is registered as attack order.
4) GPS dot appears for a moment over frozen actors sometimes.

I am trying to fix these issues but I'm having a hard time. Any help is appreciated.

deniz1a commented:
I decided to postpone this PR again. I am stuck at fixing issue #1 described above... pchote said there may be some changes to how frozen actors are rendered. So I am waiting either that to be finished or someone to show me how I could fix these problems.

deniz1a commented:
I have fixed the three bugs described above. GPS icon appearing briefly over newly rendered frozen actors is not caused by this PR, it is also present in bleed. This PR is now finished. I tested it and couldn't find any other issues.

RoosterDragon commented on OpenRA.Game/Traits/CreatesDisruptionField.cs:
```diff
@@ -0,0 +1,51 @@
+			if (lobbyFogDisabled)
+				return;
+
+			var disabled = self.TraitsImplementing<IDisable>().Any(d => d.Disabled);
```
```var disabled = self.IsDisabled();```

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -163,42 +162,105 @@ void UpdateVisibility(Actor a, ref CPos[] visible)
+		void AddDisruptionGenerator(Actor a, ref CPos[] underDisruption)
+		{
+			var cd = a.TraitOrDefault<CreatesDisruptionField>();
+			if (cd == null || a.Owner.IsAlliedWith(self.Owner) || cd.Range == WRange.Zero || !self.World.LobbyInfo.GlobalSettings.Fog)
```
Isn't the check for the disabled fog also done at the trait level? Do we need to duplicate it here?

RoosterDragon commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -179,7 +186,27 @@ public void Tick(Actor self)
 		{
 			return world.ScreenMap.FrozenActorsInBox(owner, wr.Viewport.TopLeft, wr.Viewport.BottomRight)
-				.Where(f => f.Visible)
+				.Where(f =>
```
Could you walk me through what this logic block is trying to solve? You've clearly engineered it very carefully to fix some of the issues you were seeing but it's difficult to see why each of these cases is needed. I'm wondering if there's something we could do to simply this concept.

deniz1a replied:
OK. I made that to fix the bug #1 described above. The frozen actor of the building has to be rendered at least once so that it is cached and is shown even when the building is not there anymore. This code makes sure that every frozen actor is rendered once even if it's under disruption field and then one tick passes while it is rendered. Because tick and render frequencies are different. This makes the renderables cached so that they can be shown again when the building is destroyed and disruption field is gone.

At least this is my understanding of how the engine works. Is that correct?

deniz1a replied:
OK, I unnestified the if statements. I don't think it can be simplified more than this.

pchote replied:
This sounds much more complicated than it needs to be.  A simpler solution would be to change FrozenActor to generate the sprites from an ActorPreview if the actor has been destroyed.

deniz1a replied:
How can I do that?

deniz1a commented:
I made the changes you requested. While I was doing that I also realized that there should also be a separate RenderDisruptionCircle trait to draw the range circle. And shouldn't this and RenderShroudCircle be in Mods.Common? Or even in the engine itself. Because they are part of CreatesShroud and CreatesDisruptionField traits.

deniz1a commented:
And the developing continues...
[19:54:56] <+pchote> find the thing that calls the render, and then create an actor preview
[19:55:03] <+pchote> and then render that and stash the sprites
[19:55:46] <+pchote> actually, there's a problem there
[19:55:59] <+pchote> the base ActorPreview class currently lives in Mods.Common
[19:56:07] <+pchote> it will need to move into game for you to be able to do this

deniz1a commented:
I have changed this to use ActorPreview for rendering the frozen actor. I also added some checks so that it is only rendered when the building state has changed.

abcdefg30 commented on OpenRA.Game/OpenRA.Game.csproj:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
This needs to be undone.

Phrohdoh commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -25,6 +26,7 @@ public class FrozenActorLayerInfo : ITraitInfo
 	public class FrozenActor
 	{
 		public readonly MPos[] Footprint;
+		public readonly CPos[] OccCells;
```
What is `Occ`, Occupied?

deniz1a replied:
Yes, that's for Occupied Cells. But there already is a method called OccupiedCells() so I wanted to give this a different name to avoid confusion.

Phrohdoh replied:
This sounds like duplication then.

deniz1a replied:
But that method cannot be used if the actor is null. A frozen actor can exist even after its actor is destroyed. The actor's occupied cells have to be stored in the frozen actor so that they can be used even if the actor is null.

pchote replied:
That is what the Footprint above is for. Use that instead.

deniz1a commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -48,6 +48,7 @@ public FrozenActor(Actor self, MPos[] footprint, CellRegion footprintRegion, Shr
 			isVisibleTest = shroud.IsVisibleTest(footprintRegion);
 
 			Footprint = footprint;
+			OccCells = Shroud.GetVisOrigins(actor).ToArray();
```
Should I use "OccCells = self.OccupiesSpace.OccupiedCells().Select(c => c.First).ToArray();" instead of this line? Do I need the checks in GetVisOrigins()?

deniz1a commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -383,13 +453,25 @@ public bool IsVisible(Actor a)
 		public bool IsTargetable(Actor a)
 		{
+			IEnumerable<CPos> vis = GetVisOrigins(a);
+			bool concealed = vis.All(c => generatedDisruptionCount[c] > 0 || generatedShroudCount[c] > 0);
```
pchote: But I'm using occupied cells here to determine if a building is targetable. Should I use footprint here too? I have to use the same condition everywhere.

pchote replied:
This code deals with regular actors (not FrozenActors), so I'm not sure what you're asking.

deniz1a replied:
But the checks for regular actors and frozen actors have to be the same. Otherwise there are frozen actors that are visible but not auto-targetable. So either I have to check for footprints here or add the occupied cells to frozen actor.

pchote replied:
If this is for targeting, then your "occupied cells" in the frozen actor need to instead store and deal with the targetable positions returned by the actor.  These are conceptually different to the occupied cells, even though they may (for the current set of traits) be derived from the same data.

pchote replied:
Although i'm still not convinced about all of this.  It sounds like you have uncovered an existing bug that must already exist in certain situations in the current code -- separating out the bugfix from your generation code would make this much easier to understand and review.

pchote commented on OpenRA.Game/Orders/UnitOrderGenerator.cs:
```diff
@@ -29,7 +29,8 @@ public IEnumerable<Order> Order(World world, CPos xy, MouseInput mi)
 				var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mi)
-					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures))
+					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures)
+						&& !world.RenderPlayer.Shroud.IsUnderDisruptionField(a.OccCells))
```
Requiring an explicit check for IsUnderDisruptionField here is a sign that the existing code is bogus.  This code (from before your changes) should have been filtering out FrozenActors that aren't visible, and your changes should have just worked (because it should be making FrozenActors.Visible -> false).  Please fix this properly by adding a filter on Visible instead of hacking in a disruption check in these places.

pchote commented on OpenRA.Game/Traits/CreatesShroud.cs:
```diff
@@ -21,20 +22,21 @@ public class CreatesShroudInfo : ITraitInfo
 
 	public class CreatesShroud : ITick, ISync
 	{
+		public WRange Range { get { return cachedDisabled ? WRange.Zero : info.Range; } }
```
Why did you move this line?

deniz1a replied:
Because Phrohdoh told me to.

penev92 replied:
He does have a point. Properties usually go before methods.

pchote commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -64,6 +65,8 @@ public FrozenActor(Actor self, MPos[] footprint, CellRegion footprintRegion, Shr
 		int flashTicks;
 		IRenderable[] renderables = NoRenderables;
 		bool needRenderables;
+		DamageState DamageStateCached;
```
These are private, so should the naming needs to be camelCase.

pchote commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -98,23 +101,22 @@ public void Flash()
-					renderables = actor.Render(wr).ToArray();
-					IsRendering = false;
-				}
+				IsRendering = true;
```
Please only use the ActorPreview if `actor.Destroyed` is true.  Using it for a living actor breaks the feature where the actor is frozen with its last visible frame (including muzzle flashes, turret position, etc).

pchote commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -98,23 +101,22 @@ public void Flash()
 		public IEnumerable<IRenderable> Render(WorldRenderer wr)
 		{
-			if (needRenderables)
+			if (needRenderables && (renderables == NoRenderables || DamageStateCached != DamageState || OwnerCached != Owner))
```
These extra checks don't look right to me.  if needRenderables is true, then I would expect it to generate renderables.  If there is some condition where you don't want it to generate renderables, then those conditions should be applied at the point that variable is set.

pchote commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -179,7 +181,7 @@ public void Tick(Actor self)
 		{
 			return world.ScreenMap.FrozenActorsInBox(owner, wr.Viewport.TopLeft, wr.Viewport.BottomRight)
-				.Where(f => f.Visible)
+				.Where(f => f.Visible && !owner.Shroud.IsUnderDisruptionField(f.OccCells))
```
Like I mentioned above, your disruption checks should be built into `f.Visible` instead of adding IsUnderDisruptionField everywhere.

deniz1a commented:
This PR depends on #8084.

penev92 commented:
I am closing this in an attempt to clean up the PR queue a bit. Feel free to reopen it once the conceptual problems are cleared up.

deniz1a commented:
This apparently depends on satellite code being fixed to handle visibility and stuff. Related: #8203
